### PR TITLE
Mylin/1617 support spatial profile line polyline

### LINKED ISSUE #1617: Expand the features in the spatial profiler: line region/averaging width
Expand the features in the spatial profiler widget including:

- [ ] a region dropdown menu to select cursor or a line region
- [ ] averaging width (in pixel) in the settings

### PR DESCRIPTION
**Pull request by acdo2002**
Fixed #1617 
The corresponding backend [PR1020](https://github.com/CARTAvis/carta-backend/pull/1020)

### CODE REVIEW

**Comment by kswang1029** [comment]
I have had a preview of this branch and found few issues that need to be addressed:
1. when the region is point or cursor, the x label of the profile plot is "X coordinate" or "Y coordinate" (with an implicit unit of "pixel". When the region is line (not polyline), the x label should be revised to offset with a unit (eg Offset (arcsec). The actual unit (arcsec, arcmin, or degree) is determined by the angular size of the line region and the backend should already have a handle on this based on the implementation of pv generator. When the region is polyline, the x label should be "Distance" with a unit. I suspect we need to update protobuf message SPATIAL_PROFILE_DATA. 
2. If we update the line or polyline region and before the new results come in, we need to have visual indication that the backend is working on it. Similar to region spectral profilers, we should set the old profile dimmer.
3. the red vertical line should disappear when the region type is line and polyline. (as a future enhancement, we should display a marker along the line or polyline region and sync its movement with the red vertical line in the spatial profile plot)
4. if the active line region is deleted, the spatial profiler has an odd behavior displaying the new active region.

https://user-images.githubusercontent.com/20819712/163362347-1ed4616d-c794-492b-b1c2-4d040ac6bd2e.mov

5. the title of the spatial profiler widget should be more dynamic based on what is being displayed. Before the support of line and polyline region, it is like "X Profile: Cursor". Now this should be like "Profile: region 2" when region 2 is a line region or a polyline region. 
6. some UI elements in the smoothing tab of the settings dialog are not rendered properly.
Screen Shot 2022-04-14 at 18 04 45

**Comment by acdo2002** [comment]
@kswang1029 
I have shown the current frontend to Tien-Hao and already discussed with Tien-Hao.
1. may need the modification of ICD message, I leave it later.
2. may need the modification of ICD message, I leave it later.
3. Fixed, I have let the red vertical line to be transparent if the region is line/polyline. (the enhancement please issue it in the future)
4. It is because once the region (no matter the region is line/polyline) is deleted, the cursor will show the previous cursor spatial profile data. If you move the cursor (slightly around the region) before you delete the region, the cursor will show a relatively correct value. We think this is another issue.
5. Fixed the coordinate string, now it will show "Profile: Region # number" in the spatial profiler widget
6. Fixed the scss.

**Comment by kswang1029** [comment]
> 4. It is because once the region (no matter the region is line/polyline) is deleted, the cursor will show the previous cursor spatial profile data. If you move the cursor (slightly around the region) before you delete the region, the cursor will show a relatively correct value. We think this is another issue.

issue filed at https://github.com/CARTAvis/carta-frontend/issues/1831

**Comment by kswang1029** [comment]
@acdo2002 the displayed unit in the cursor info text string needs to be sync'd to the actual unit of the line/polyline region. Also the WCS part needs to be removed when it is line or polyline region. 
Screen Shot 2022-04-26 at 13 57 48

**Comment by kswang1029** [comment]
@acdo2002 when it is line or polyline profile being displayed, the header of the exported tsv file needs to be updated and include the line or polyline information. At the moment it is still a point region. 
Screen Shot 2022-04-26 at 14 10 36

**Comment by kswang1029** [comment]
@acdo2002 spotted a frontend crash. If we explicitly set the image dropdown to an image (not "Active") and explicitly set the region dropdown to a line region (not "Active"), then if we delete the region, frontend will crash.
Screen Shot 2022-04-26 at 14 38 31

**Comment by kswang1029** [comment]
@TienHao as we have a new “compute” tab in the settings dialog, please add a placeholder for the in-app help.

**Comment by kswang1029** [comment]
@TienHao Please help to refine the cursor info:
1. when the region is line or polyline, we don't need to show world coordinate in the cursor info
2. please switch to scientific notation to show distance or offset. %.5e should be good enough. 
3. Replace "Image" to "Offset" (for line region) or "Distance" (for polyline region). 

Screen Shot 2022-04-27 at 11 41 38

**Comment by kswang1029** [comment]
@TienHao The tsv header and file name need another tweak. When the profile is from line or polyline region, we should say "Region Profile" instead of "X profile" in the first line of the tsv file. Similar for the file name of the tsv.
Screen Shot 2022-04-27 at 11 55 05

**Comment by kswang1029** [comment]
@TienHao please enable auto y-scale when zooming for a line or polyline profile, just like point or cursor region profile.

**Comment by kswang1029** [comment]
@TienHao The title of the settings dialog needs a tweak too. I suggest we unify it to "Spatial Profile Settings"

In addition, the show mean/rms function does not work for line/polyline profile. 
Screen Shot 2022-04-27 at 13 26 18

**Comment by kswang1029** [comment]
@YuHsuan-Hwang @veggiesaurus I have had brief tests during the draft PR stage. Now it is ready for code review. I will continue my tests before approving.

**Review by kswang1029 — APPROVED** [review]
@TienHao @acdo2002 this works nicely. 👍

File: src/components/SpatialProfiler/SpatialProfilerComponent.tsx [hunk 1]
```diff
@@ -384,10 +437,11 @@ export class SpatialProfilerComponent extends React.Component<WidgetProps> {
         }
 
         const isXProfile = widgetStore.isXProfile;
+        const xLabel = this.lineAxis ? `${this.lineAxis.label} (${this.lineAxis.unit ?? ""})` : `${isXProfile ? "X" : "Y"} coordinate`;
         const imageName = appStore.activeFrame ? appStore.activeFrame.filename : undefined;
-        const plotName = `${isXProfile ? "X" : "Y"} profile`;
+        const plotName = `${this.lineAxis ? "" : isXProfile ? "X" : "Y"} profile`;
```
Comment (YuHsuan-Hwang) [inline src/components/SpatialProfiler/SpatialProfilerComponent.tsx]: remove extra space for line regions in filename
image

File: src/stores/Frame/RegionStore.ts [hunk 2]
```diff
@@ -239,6 +239,8 @@ export class RegionStore {
         switch (this.regionType) {
             case CARTA.RegionType.POINT:
                 return `Point (pixel) [${center}]`;
+            case CARTA.RegionType.LINE:
+                return `Line (pixel) [[${this.controlPoints[0].x}, ${this.controlPoints[0].y}], [${this.controlPoints[1].x}, ${this.controlPoints[1].y}]]`;
```
Comment (YuHsuan-Hwang) [inline src/stores/Frame/RegionStore.ts]: Shall we fix the precision to 6 and add unit `pix`? Shall we show "[Invalid]" if the value does not exist?

File: src/stores/Frame/RegionStore.ts [hunk 3]
```diff
@@ -250,6 +252,13 @@ export class RegionStore {
                     polygonProperties += index !== this.controlPoints.length - 1 ? ", " : "]";
                 });
                 return polygonProperties;
+            case CARTA.RegionType.POLYLINE:
+                let polylineProperties = "Polyline (pixel) [";
+                this.controlPoints.forEach((point, index) => {
+                    polylineProperties += isFinite(point.x) && isFinite(point.y) ? `[${toFixed(point.x, 6)}, ${toFixed(point.y, 6)}]` : "[Invalid]";
```
Comment (YuHsuan-Hwang) [inline src/stores/Frame/RegionStore.ts]: Shall we add unit `pix`?

**Review by YuHsuan-Hwang — CHANGES_REQUESTED** [review]
The changes look good to me. only very minor issues.

File: src/components/SpatialProfiler/SpatialProfilerComponent.tsx (hunk 1 above)
Comment (acdo2002) [reply src/components/SpatialProfiler/SpatialProfilerComponent.tsx]: Fixed!

File: src/stores/Frame/RegionStore.ts (hunk 2 above)
Comment (acdo2002) [reply src/stores/Frame/RegionStore.ts]: I add the lineProperties, making the format similar to polygonProperties & polylineProperties.

File: src/stores/Frame/RegionStore.ts (hunk 3 above)
Comment (acdo2002) [reply src/stores/Frame/RegionStore.ts]: Fixed!

**Comment by acdo2002** [comment]
@YuHsuan-Hwang I try my best to solve the merge conflicts (as well as solve the minor issue you suggested), could you check whether this branch is working fine or not?
At least this branch can build successfully on my MacOS 11 laptop, I will try to build on my Ubuntu laptop in the weekend.

**Review by YuHsuan-Hwang — APPROVED** [review]
> @YuHsuan-Hwang I try my best to solve the merge conflicts (as well as solve the minor issue you suggested), could you check whether this branch is working fine or not?
At least this branch can build successfully on my MacOS 11 laptop, I will try to build on my Ubuntu laptop in the weekend.

@acdo2002 looks fine to me 👍

**Comment by acdo2002** [comment]
I have tried to build it on ubuntu 20.04, it looks fine.

**Pull request merged**